packages/schema/src/language-server/zmodel-workspace-manager.ts

```diff
@@ -1,21 +1,115 @@
-import { DefaultWorkspaceManager, LangiumDocument } from 'langium';
-import path from 'path';
-import { WorkspaceFolder } from 'vscode-languageserver';
-import { URI } from 'vscode-uri';
-import { STD_LIB_MODULE_NAME } from './constants';
+import { isPlugin, Model } from '@zenstackhq/language/ast';
+import { getLiteral } from '@zenstackhq/sdk';
+import { DefaultWorkspaceManager, interruptAndCheck, LangiumDocument } from 'langium';
+import path, { resolve } from 'path';
+import { CancellationToken, WorkspaceFolder } from 'vscode-languageserver';
+import { URI, Utils } from 'vscode-uri';
+import { PLUGIN_MODULE_NAME, STD_LIB_MODULE_NAME } from './constants';
 
 /**
  * Custom Langium WorkspaceManager implementation which automatically loads stdlib.zmodel
  */
 export default class ZModelWorkspaceManager extends DefaultWorkspaceManager {
+    public pluginModels = new Set<string>();
+
     protected async loadAdditionalDocuments(
         _folders: WorkspaceFolder[],
         _collector: (document: LangiumDocument) => void
     ): Promise<void> {
         await super.loadAdditionalDocuments(_folders, _collector);
         const stdLibUri = URI.file(path.join(__dirname, '../res', STD_LIB_MODULE_NAME));
+
+        resolve(__dirname, '../res', STD_LIB_MODULE_NAME);
         console.log(`Adding stdlib document from ${stdLibUri}`);
         const stdlib = this.langiumDocuments.getOrCreateDocument(stdLibUri);
         _collector(stdlib);
     }
+
+    override async initializeWorkspace(
+        folders: WorkspaceFolder[],
+        cancelToken = CancellationToken.None
+    ): Promise<void> {
+        const fileExtensions = this.serviceRegistry.all.flatMap((e) => e.LanguageMetaData.fileExtensions);
+        const documents: LangiumDocument[] = [];
+        const collector = (document: LangiumDocument) => {
+            documents.push(document);
+            if (!this.langiumDocuments.hasDocument(document.uri)) {
+                this.langiumDocuments.addDocument(document);
+            }
+        };
+        // Even though we don't await the initialization of the workspace manager,
+        // we can still assume that all library documents and file documents are loaded by the time we start building documents.
+        // The mutex prevents anything from performing a workspace build until we check the cancellation token
```

Comment:
What does this "mutex" mean here?

Reply:
It's copied from the base class `DefaultWorkspaceManager`.  I think it means that the `initializeWorkspace` function itself is actually running in the `mutex.lock`.  Anyway, I think I will just remove it